pysph/tools/geometry_stl.pyx

```
@@ -0,0 +1,190 @@
+                                 np.arange(st_min[1], st_max[1], h))
+
+    s_mesh, t_mesh = s_mesh.ravel(), t_mesh.ravel()
+    mask = np.empty(len(s_mesh), dtype='bool')
+
+    for i in range(len(s_mesh)):
+        mask[i] = in_triangle(s_mesh[i], t_mesh[i],
+                              st[0, 0], st[0, 1],
+                              st[1, 0], st[1, 1],
+                              st[2, 0], st[2, 1])
+
+    s_mesh, t_mesh = s_mesh[mask], t_mesh[mask]
+    st_mesh = np.vstack([s_mesh, t_mesh])
+
+    """
+    Final mesh coordinates generated from parameters s and t
+    """
+    result = np.dot(s_mesh.reshape(-1, 1), a.reshape(-1, 1).T) + \
+        np.dot(t_mesh.reshape(-1, 1), b.reshape(-1, 1).T) + p
+    return result[:, 0], result[:, 1], result[:, 2]
+
+
+def get_neighbouring_particles(pa_src, pa_dst, radius_scale):
+    """
+    Parameters
+    ----------
+    pa_src : Source particle array 
+    pa_dst : Destination particle array
+    """
+    pa_list = [pa_dst, pa_src]
+    nps = nnps.LinkedListNNPS(dim=3, particles=pa_list,
+                              radius_scale=radius_scale)
+
+    nps.set_context(src_index=1, dst_index=0)
+
+    n = len(pa_dst.x)
+    nbrs = UIntArray()
+    grid_set = set()
+    for i in range(n):
+        nps.get_nearest_particles(1, 0, i, nbrs)
+        neighbours = list(nbrs.get_npy_array())
+        for neighbour in neighbours:
+            grid_set.add(neighbour)
+
+    idx = list(grid_set)
+
+    return pa_src.x[idx], pa_src.y[idx], pa_src.z[idx]
+
+
+def get_stl_mesh(stl_fname, dx_triangle):
+    m = mesh.Mesh.from_file(stl_fname)
+    x_list, y_list, z_list = [], [], []
+    for i in range(len(m.vectors)):
+        try:
+            x1, y1, z1 = fill_triangle(m.vectors[i], dx_triangle)
+            x2, y2, z2 = get_triangle_sides(m.vectors[i], dx_triangle)
+            x_list.append(np.r_[x1, x2])
+            y_list.append(np.r_[y1, y2])
+            z_list.append(np.r_[z1, z2])
+        except Exception as e:
```

Don't catch all exceptions, just catch ValueError, in fact, create a ZeroTriangleError exception of your own and catch that.

---

This stl mesh will have many coincident points, is that correct?

---

Indeed.